tee-launcher/launcher.py

```
@@ -141,62 +296,113 @@ def main():
-
-        # Avoid getting rate-limited by the server.
-        time.sleep(0.1)
+        manifest_url = f"https://{docker_image.registry()}/v2/{docker_image.name()}/manifests/{tag}"
+        headers = {
+            "Accept": "application/vnd.docker.distribution.manifest.v2+json",
+            "Authorization": f"Bearer {token}"
+        }
+        try:
+            manifest_resp = request_until_success(
+                url=manifest_url,
+                headers=headers,
+                rpc_request_interval_secs=rpc_request_interval_secs,
+                rpc_request_timeout_secs=rpc_request_interval_secs,
+                rpc_max_attempts=rpc_max_attempts)
+            manifest = manifest_resp.json()
+            match manifest['mediaType']:
+                case 'application/vnd.oci.image.index.v1+json':
+                    # Multi-platform manifest; we scan for amd64/linux images and add them to `tags`
+                    for image_manifest in manifest.get('manifests', []):
+                        platform = image_manifest.get('platform', [])
+                        if platform.get(
+                                'architecture') == 'amd64' and platform.get(
+                                    'os') == 'linux':
+                            tags.append(image_manifest['digest'])
+                case 'application/vnd.docker.distribution.manifest.v2+json' | \
+                     'application/vnd.oci.image.manifest.v1+json':
+                    config_digest = manifest['config']['digest']
+                    if config_digest == docker_image.digest:
+                        return manifest_resp.headers['Docker-Content-Digest']
+        except:
+            print(
+                "[Warning] Exceeded number of maximum RPC requests for any given attempt. Will continue in the hopes of finding the matching image hash among remaining tags"
+            )
+            # Q: Do we expect all requests to succeed?
```

do we?

---

don't know

---

Feels weird to keep looping if we have failed the retry loops. If we hit rate limits, wouldn't it be better to use a pretty long exponential backoff? And if we still fail I think we should raise an exception here and let the launcher fail explicitly instead of hanging and continue sending failing requests.

---

> And if we still fail I think we should raise an exception here and let the launcher fail explicitly instead of hanging and continue sending failing requests.

IIUC, then we don't necessarily require these requests to succeed. Seems to me like we are searching for a matching hash https://github.com/near/mpc/blob/3610a24b1434a91dfc46f642cd40e3372d2a8c20/tee-launcher/launcher.py#L393

Before this PR, we ignored any tag for which we got a response different from 200.
This meant that if we hit a rate limit at the hash of interest (or any other error that was recoverable), the launcher would not retry and eventually fail, since we would skip the tag of interest and probably not find it anymore.

With this PR, we try to give us the best chances of success. This might lead to wasting some resources, but I don't understand this endpoint enough to judge that. (I would guess @barakeinav1 has more insights here)

For example, if we request a tag that does not exist, will we get a `400` or a `404`? In such a case, we should probably just skip the tag. But given my limited understanding of this endpoint and how it is expected to behave, I thought it would be better to err on the side of caution and just retry up to the allotted limit.

If it turns out to be bad performance wise, we can always increase the timeout and decrease the maximum number of attempts through the environment variables.